File: 16.Software-Design-Patterns/example/Creational/Multiton/Multiton.js
class MultitonJs {
    static instances = new Map();
  
    constructor(id) {
      this.id = id;
    }
  
    static getInstance(key) {
      if (!this.instances.has(key)) {
        this.instances.set(key, new MultitonJs(key));
      }
      return this.instances.get(key);
    }
  
    log(message) {
      console.log(`Logger (${this.id}): ${message}`);
    }
  }
  
  const example__1 = MultitonJs.getInstance("app");
  example__1.log("This is the first logger instance");
  
  const example__2 = MultitonJs.getInstance("api");
  example__2.log("This is the second logger instance");
  
  console.log(example__1 === example__2);
  

File: 16.Software-Design-Patterns/example/Creational/Multiton/Multiton.java
import java.util.HashMap;
import java.util.Map;

public class Multiton<T> {
    private static final Map<String, Multiton<?>> instances = new HashMap<>();
    private final String id;

    private Multiton(String id) {
        this.id = id;
    }

    public static synchronized Multiton<?> getInstance(String key) {
        if (!instances.containsKey(key)) {
            instances.put(key, new Multiton<>(key));
        }
        return instances.get(key);
    }

    public void log(String message) {
        System.out.println("Logger: " + message);
    }

    public static void main(String[] args) {
        Multiton<?> example_1 = Multiton.getInstance("app");
        example_1.log("This is the first logger instance");

        Multiton<?> example_2 = Multiton.getInstance("api");
        example_2.log("This is the second logger instance");

        System.out.println(example_1 == example_2); 
    }
}
